rfcs/0000-compiled-runtime.md

```diff
@@ -0,0 +1,417 @@
+import { css, keyframes } from '@compiled/react';
+
+const fadeIn = keyframes({
+  to: { opacity: 1 },
+});
+
+const styles = css({
+  animationName: fadeIn,
+  animationDuration: '2s',
+});
+
+// 👇 transforms to
+
+const styles = i([
+  '_5sagymdr _5sagymdr',
+  [
+    '._j7hqa2t1{animation-name:kfwl3rt}',
+    '@keyframes kfwl3rt{to{opacity:1}}',
+    '._5sagymdr{animation-duration:2s}',
+  ],
+]);
+```
+
+**Considerations**
+
+- Could there be an alternate way to implement this that leans on similar patterns as the previous API?
+
+### CSS Prop
+
+CSS prop would now be wired up at runtime for both application of class names and style insertion. A component would be created for this. This component would use the `ax` function to build a `className` as well as the `CC` and `CS` components to handle SSR/client side style insertion. All other APIs would use them too.
+
+```jsx
+/** @jsxImportSource @compiled/react */
+
+const redColorStyles = css({ color: 'red' });
+const hiddenStyles = css({ display: 'none' });
+const largeTextStyles = css({ fontSize: 50 });
+
+function Comp({ isRed, isHidden, isLarge }) {
+  return (
+    <div css={[isHidden ? hiddenStyles : [isLarge && largeTextStyles, isRed && redColorStyles]]} />
+  );
+}
+
+// 👇 transforms to
+
+const redColorStyles = i(['_syaz5scu', ['._syaz5scu{color:red}']]);
+const hiddenStyles = i(['_1e0cglyw', ['._1e0cglyw{display:none}']]);
+const largeTextStyles = i(['_1wyb1sen', ['._1wyb1sen{font-size:50}']]);
+
+function Comp({ isRed, isHidden, isLarge }) {
+  return (
+    <div css={[isHidden ? hiddenStyles : [isLarge && largeTextStyles, isRed && redColorStyles]]} />
+  );
+}
+```
+
+Style extraction then becomes a matter of pointing to the extract runtime and stripping out the styles from their arrays. The extract runtime would remove all style insertion behavior.
+
+```diff
+-/** @jsxImportSource @compiled/react */
++/** @jsxImportSource @compiled/react/runtime */
+
+-const redColorStyles = i(['_syaz5scu', ['._syaz5scu{color:red}']]);
+-const hiddenStyles = i(['_1e0cglyw', ['._1e0cglyw{display:none}']]);
+-const largeTextStyles = i(['_1wyb1sen', ['._1wyb1sen{font-size:50}']]);
++const redColorStyles = i(['_syaz5scu']);
++const hiddenStyles = i(['_1e0cglyw']);
++const largeTextStyles = i(['_1wyb1sen']);
+
+function Comp({ isRed, isHidden, isLarge }) {
+  return (
+    <div
+      css={[
+        isHidden
+          ? hiddenStyles
+          : [isLarge && largeTextStyles, isRed && redColorStyles],
+      ]}
+    />
+  );
+}
+```
+
+### Styled
+
+Styled components would build on the same foundations as both the `css` func & prop, still leaning on a runtime to do the application of class names and style insertion.
+
+```jsx
+import { css, styled } from '@compiled/react';
+
+const Comp = styled.div`
+  color: ${(props) => props.color};
+  ${(props) =>
+    props.isHidden
+      ? css({ display: 'none' })
+      : [props.isLarge && css({ fontSize: 50 }), props.isRed && css({ color: 'red' })]}
+`;
```

Comment:
I wonder if it's worth first converting this to something like:
`<div css={...css(stringTemplateContent) }>` 

Would this reduce code needed to handle the `styled` API?

Reply:
Not really, styled needs to internally forward ref and construct the JSX element that was wrapped (either native or composed) - all that to say it would be easier to just write the small abstraction to handle it.

In one way you can imagine every time a "compiled component" is transformed into the AST - that's just where the abstraction would be used instead. So while we currently don't really count them as apart of the bundle size, it might end up reducing it even further by creating the abstractions even though the library "looks" bigger.

Reply:
That said thinking about this further realistically the component that would be backing the jsx pragma would forward ref too... so it probably just comes down value prop of every component flushing through css pragma vs. just styled, as the former means turning the jsx pragma on and that could be a bit... finicky? Could be worth fleshing this out further though!